### ISSUE TITLE
Composer lockfile and dependencies are not compatible with PHP 7.1+

### ISSUE THREAD

**Issue by justlevine** [wp-graphql/wp-graphql#2451]
## Problem
While the minimum PHP version of WPGraphQL is 7.1, the composer lockfile is built based on whatever PHP version was used for the last commit. The lockfile is _not_ compatible with PHP8.0+. Additionally, our dev-dependencies (specifically PHPUnit 9.x) requires a minimum version of PHP 7.3.

While not an issue for end-users (since there's no composer deps in the final build), this is an issue for developers who wish to run tests/linting locally, and arent using PHP v7.3 or v7.4, as well as a bit of a headache when merging commits that update composer.lock.

**Bonus:** This would also also us to use `composer install` instead of `composer update` for our GH workflows, to prevent failures caused by updated deps.

## Proposed solution
Broadly, the way to solve this is to set `config.platform.php` to an explicit PHP version.
_However_, we cannot set it to `7.1.0`, without downgrading PHPUnit to v8.x. This is a problem since the latest version of wp-graphql-test-case is [no longer compatible with PHPUnit v8.x](https://github.com/wp-graphql/wp-graphql-testcase/issues/19#issuecomment-1140140402).

As such we have 2 options:
- **NOT RECOMMENDED**: Update the minimum PHP version for the plugin to v7.3+. This is bad, since 1) it's a breaking change, 2) there's no reason to break _plugin_ compatibility for a dev issue, and 3) the [latest `develop` branch of `graphql-php`](https://github.com/webonyx/graphql-php/blob/master/composer.json), bumps its minimum PHP version to 7.4, so at some point theoretically soon we may need to bump the deps again.

- **RECOMMENDED**: Set  `config.platform.php` to a later PHP version, and leave a note to developers about the minimum requirements. While I prefer this, we would need to choose a version to support. I'd suggest skipping 7.3 and going straight to 7.4 both based on [usage stats](https://packagist.org/packages/wp-graphql/wp-graphql/php-stats#1.8), but there's not that much of a difference from a root deps perspective(see screenshot 4). I _think_ the latest version was built with v8.1 (based on the errors from php 8.1. 

Tl;dr @jasonbahl let me know what version to use and I'm happy to submit a PR.

![image](https://user-images.githubusercontent.com/29322304/180523517-895658e7-cad5-43ad-9e61-c8f25057e865.png)

## Screenshots
1. `composer install` on PHP 8.0
![image](https://user-images.githubusercontent.com/29322304/180527592-e581f2c6-7f40-427f-b7cf-a86373165735.png)

2. Installing with `config.platform.php` set to PHP 7.1
![image](https://user-images.githubusercontent.com/29322304/180530680-b8013137-4eb8-4cc6-b87a-0e5bcc910393.png)

3. Installing with `config.platform.php` set to PHP 7.2
![image](https://user-images.githubusercontent.com/29322304/180533434-7795c14a-8ae0-4542-814d-a22ad546f9ec.png)

4. Updating with `config.platform.php` set to PHP 7.3->7.4.
![image](https://user-images.githubusercontent.com/29322304/180560984-df425f67-2f29-47c9-927e-23e480171b39.png)

**Comment by jasonbahl** [comment]
What a great, detailed issue. Thanks so much @justlevine! 🙌🏻 

I like your recommendation! 

I don't have a strong opinion on 7.4, 8.0 or 8.1. 🤷🏻‍♂️ 

One thing to note, however, is that while we claim the minimum version for the plugin is PHP 7.1, we don't actually run tests on 7.1, so we're actually not even guaranteeing this is true right now. 🤔 

We should probably update the test matrix to run tests in 7.1 (if possible) or update the minimum to a PHP version we're _actually_ supporting. 🤔

**Comment by justlevine** [comment]
> I don't have a strong opinion on 7.4, 8.0 or 8.1. 🤷🏻‍♂️

In that case I think we choose 7.4, to make it easier for more devs to contribute, since that will work in later versions, but not vice versa. Will submit a PR shortly.

> One thing to note, however, is that while we claim the minimum version for the plugin is PHP 7.1, we don't actually run tests on 7.1, so we're actually not even guaranteeing this is true right now. 🤔
> We should probably update the test matrix to run tests in 7.1 (if possible) or update the minimum to a PHP version we're actually supporting. 🤔

#2450 will fix `phpcs` so its correctly sniffing for code incompatible with 7.1+. I think that might be enough until there's a reason to drop explicit support for a version (e.g. an incompatibility bug we cant fix, or when we're ready for a breaking release).

That's a lot of extra combinations to add to the build matrix, and we can also make the same argument for WP versions (we're not testing every version since 5.0). Your call 🤷‍♂️

**Comment by jasonbahl** [comment]
I just quickly tried to test running tests in PHP 7.1 with WP 5.3 as that's an [official WordPress docker image](https://hub.docker.com/layers/wordpress/library/wordpress/5.3.0-php7.1-apache/images/sha256-ba4e28d21f00f3a92a68c975ab1325ba9d199beec5c22c345f3b2e8ce25a59bf?context=explore) by running this command: 

`PHP_VERSION=7.1 WP_VERSION=5.3 composer build-test` 
then
` PHP_VERSION=7.1 WP_VERSION=5.3 composer run-test`

But there's some issues. 

The `build-test` command will require Xdebug v2.9.0 to be installed when `PHP_VERSION=7.1` (see: https://github.com/flyve-mdm/docker-environment/issues/68#issue-787535093) and we'll have to modify some composer dependencies in the run-test command to play nice with 7.1. 

It would be nice to confirm for sure that WPGraphQL does actually work well in PHP 7.1, because I can't confidently say today that it does. I assume it does, but I'm not really sure if anything has been added that would make it incompatible. 🤷🏻‍♂️

**Comment by justlevine** [comment]
PHPUnit 9 anyway requires PHP v7.3+, and  PHPUnit 8.5 still needs at least PHP v7.2+( if we wanted to revert to the older version of wp-graphql-test-case and downgrade), so even if you can build the image, you wont be able to do anything with it.

Tangentially related, we also probably [dont need pcov/clobber](https://github.com/wp-graphql/wp-graphql/blob/4519bb96811e8a4fbcadba75282c93ac60466ae0/docker/testing.entrypoint.sh#L95) since its built into PHPUnit 8+.

**Comment by justlevine** [comment]
Had to opt for `PHP 7.3`, as `codeception/stub` v4 breaks with (only) 7.4 support, and thats the lowest php version we can test against. There's no difference in the versions for our dev dependencies between 7.4->7.3, so I say we leave it like that, and then bump the minimum version across the board to 7.4 once its truly necessary.
